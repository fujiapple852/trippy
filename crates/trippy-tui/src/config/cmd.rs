use crate::config::binding::TuiCommandItem;
use crate::config::theme::TuiThemeItem;
use crate::config::{
    AddressFamilyConfig, AddressMode, AsMode, DnsResolveMethodConfig, GeoIpMode, IcmpExtensionMode,
    LogFormat, LogSpanEvents, Mode, MultipathStrategyConfig, ProtocolConfig, TuiColor,
    TuiKeyBinding,
};
use anyhow::anyhow;
use clap::builder::Styles;
use clap::Parser;
use clap_complete::Shell;
use std::net::IpAddr;
use std::str::FromStr;
use std::time::Duration;

/// Trace a route to a host and record statistics
#[expect(clippy::doc_markdown)]
#[derive(Parser, Debug)]
#[command(name = "trip", author, version, about, long_about = None, arg_required_else_help(true), styles=Styles::styled())]
pub struct Args {
    /// A space delimited list of hostnames and IPs to trace
    #[arg(required_unless_present_any(["print_tui_theme_items", "print_tui_binding_commands", "print_config_template", "generate", "generate_man", "print_locales"]))]
    pub targets: Vec<String>,

    /// Config file
    #[arg(value_enum, short = 'c', long, value_hint = clap::ValueHint::FilePath)]
    pub config_file: Option<String>,

    /// Output mode [default: tui]
    #[arg(value_enum, short = 'm', long)]
    pub mode: Option<Mode>,

    /// Trace without requiring elevated privileges on supported platforms [default: false]
    #[arg(short = 'u', long)]
    pub unprivileged: bool,

    /// Tracing protocol [default: icmp]
    #[arg(value_enum, short = 'p', long)]
    pub protocol: Option<ProtocolConfig>,

    /// Trace using the UDP protocol
    #[arg(
        long,
        conflicts_with = "protocol",
        conflicts_with = "tcp",
        conflicts_with = "icmp"
    )]
    pub udp: bool,

    /// Trace using the TCP protocol
    #[arg(
        long,
        conflicts_with = "protocol",
        conflicts_with = "udp",
        conflicts_with = "icmp"
    )]
    pub tcp: bool,

    /// Trace using the ICMP protocol
    #[arg(
        long,
        conflicts_with = "protocol",
        conflicts_with = "udp",
        conflicts_with = "tcp"
    )]
    pub icmp: bool,

    /// The address family [default: system]
    #[arg(value_enum, short = 'F', long)]
    pub addr_family: Option<AddressFamilyConfig>,

    /// Use IPv4 only
    #[arg(
        short = '4',
        long,
        conflicts_with = "ipv6",
        conflicts_with = "addr_family"
    )]
    pub ipv4: bool,

    /// Use IPv6 only
    #[arg(
        short = '6',
        long,
        conflicts_with = "ipv4",
        conflicts_with = "addr_family"
    )]
    pub ipv6: bool,

    /// The target port (TCP & UDP only) [default: 80]
    #[arg(long, short = 'P')]
    pub target_port: Option<u16>,

    /// The source port (TCP & UDP only) [default: auto]
    #[arg(long, short = 'S')]
    pub source_port: Option<u16>,

    /// The source IP address [default: auto]
    #[arg(short = 'A', long, value_parser = parse_addr, conflicts_with = "interface")]
    pub source_address: Option<IpAddr>,

    /// The network interface [default: auto]
    #[arg(short = 'I', long)]
    pub interface: Option<String>,

    /// The minimum duration of every round [default: 1s]
    #[arg(short = 'i', long, value_parser = parse_duration)]
    pub min_round_duration: Option<Duration>,

    /// The maximum duration of every round [default: 1s]
    #[arg(short = 'T', long, value_parser = parse_duration)]
    pub max_round_duration: Option<Duration>,

    /// The period of time to wait for additional ICMP responses after the target has responded
    /// [default: 100ms]
    #[arg(short = 'g', long, value_parser = parse_duration)]
    pub grace_duration: Option<Duration>,

    /// The initial sequence number [default: 33434]
    #[arg(long)]
    pub initial_sequence: Option<u16>,

    /// The Equal-cost Multi-Path routing strategy (UDP only) [default: classic]
    #[arg(value_enum, short = 'R', long)]
    pub multipath_strategy: Option<MultipathStrategyConfig>,

    /// The maximum number of in-flight ICMP echo requests [default: 24]
    #[arg(short = 'U', long)]
    pub max_inflight: Option<u8>,

    /// The TTL to start from [default: 1]
    #[arg(short = 'f', long)]
    pub first_ttl: Option<u8>,

    /// The maximum number of TTL hops [default: 64]
    #[arg(short = 't', long)]
    pub max_ttl: Option<u8>,

    /// The size of IP packet to send (IP header + ICMP header + payload) [default: 84]
    #[arg(long)]
    pub packet_size: Option<u16>,

    /// The repeating pattern in the payload of the ICMP packet [default: 0]
    #[arg(long)]
    pub payload_pattern: Option<u8>,

    /// The TOS (i.e. DSCP+ECN) IP header value (IPv4 only) [default: 0]
    #[arg(short = 'Q', long)]
    pub tos: Option<u8>,

    /// Parse ICMP extensions
    #[arg(short = 'e', long)]
    pub icmp_extensions: bool,

    /// The socket read timeout [default: 10ms]
    #[arg(long, value_parser = parse_duration)]
    pub read_timeout: Option<Duration>,

    /// How to perform DNS queries [default: system]
    #[arg(value_enum, short = 'r', long)]
    pub dns_resolve_method: Option<DnsResolveMethodConfig>,

    /// Trace to all IPs resolved from DNS lookup [default: false]
    #[arg(short = 'y', long)]
    pub dns_resolve_all: bool,

    /// The maximum time to wait to perform DNS queries [default: 5s]
    #[arg(long, value_parser = parse_duration)]
    pub dns_timeout: Option<Duration>,

    /// The time-to-live (TTL) of DNS entries [default: 300s]
    #[arg(long, value_parser = parse_duration)]
    pub dns_ttl: Option<Duration>,

    /// Lookup autonomous system (AS) information during DNS queries [default: false]
    #[arg(long, short = 'z')]
    pub dns_lookup_as_info: bool,

    /// The maximum number of samples to record per hop [default: 256]
    #[arg(long, short = 's')]
    pub max_samples: Option<usize>,

    /// The maximum number of flows to record [default: 64]
    #[arg(long)]
    pub max_flows: Option<usize>,

    /// How to render addresses [default: host]
    #[arg(value_enum, short = 'a', long)]
    pub tui_address_mode: Option<AddressMode>,

    /// How to render autonomous system (AS) information [default: asn]
    #[arg(value_enum, long)]
    pub tui_as_mode: Option<AsMode>,

    /// Custom columns to be displayed in the TUI hops table [default: holsravbwdt]
    #[arg(long)]
    pub tui_custom_columns: Option<String>,

    /// How to render ICMP extensions [default: off]
    #[arg(value_enum, long)]
    pub tui_icmp_extension_mode: Option<IcmpExtensionMode>,

    /// How to render GeoIp information [default: short]
    #[arg(value_enum, long)]
    pub tui_geoip_mode: Option<GeoIpMode>,

    /// The maximum number of addresses to show per hop [default: auto]
    #[arg(short = 'M', long)]
    pub tui_max_addrs: Option<u8>,

    /// Preserve the screen on exit [default: false]
    #[arg(long)]
    pub tui_preserve_screen: bool,

    /// The TUI refresh rate [default: 100ms]
    #[arg(long, value_parser = parse_duration)]
    pub tui_refresh_rate: Option<Duration>,

    /// The maximum ttl of hops which will be masked for privacy [default: none]
    ///
    /// If set, the source IP address and hostname will also be hidden.
    #[arg(long)]
    pub tui_privacy_max_ttl: Option<u8>,

    /// The locale to use for the TUI [default: auto]
    #[arg(long)]
    pub tui_locale: Option<String>,

    /// The timezone to use for the TUI [default: auto]
    ///
    /// The timezone must be a valid IANA timezone identifier.
    #[arg(long)]
    pub tui_timezone: Option<String>,

    /// The TUI theme colors [item=color,item=color,..]
    #[arg(long, value_delimiter(','), value_parser = parse_tui_theme_color_value)]
    pub tui_theme_colors: Vec<(TuiThemeItem, TuiColor)>,

    /// Print all TUI theme items and exit
    #[arg(long)]
    pub print_tui_theme_items: bool,

    /// The TUI key bindings [command=key,command=key,..]
    #[arg(long, value_delimiter(','), value_parser = parse_tui_binding_value)]
    pub tui_key_bindings: Vec<(TuiCommandItem, TuiKeyBinding)>,

    /// Print all TUI commands that can be bound and exit
    #[arg(long)]
    pub print_tui_binding_commands: bool,

    /// The number of report cycles to run [default: 10]
    #[arg(short = 'C', long)]
    pub report_cycles: Option<usize>,

    /// The supported MaxMind or IPinfo GeoIp mmdb file
    #[arg(short = 'G', long, value_hint = clap::ValueHint::FilePath)]
    pub geoip_mmdb_file: Option<String>,

    /// Generate shell completion
    #[arg(long)]
    pub generate: Option<Shell>,

    /// Generate ROFF man page
    #[arg(long)]
    pub generate_man: bool,

    /// Print a template toml config file and exit
    #[arg(long)]
    pub print_config_template: bool,

    /// Print all available TUI locales and exit
    #[arg(long)]
    pub print_locales: bool,

    /// The debug log format [default: pretty]
    #[arg(long)]
    pub log_format: Option<LogFormat>,

    /// The debug log filter [default: trippy=debug]
    #[arg(long)]
    pub log_filter: Option<String>,

    /// The debug log format [default: off]
    #[arg(long)]
    pub log_span_events: Option<LogSpanEvents>,

    /// Enable verbose debug logging
    #[arg(short = 'v', long, default_value_t = false)]
    pub verbose: bool,
}

fn parse_tui_theme_color_value(value: &str) -> anyhow::Result<(TuiThemeItem, TuiColor)> {
    let pos = value
        .find('=')
        .ok_or_else(|| anyhow!("invalid theme value: expected format `item=value`"))?;
    let item = TuiThemeItem::try_from(&value[..pos])?;
    let color = TuiColor::try_from(&value[pos + 1..])?;
    Ok((item, color))
}

fn parse_tui_binding_value(value: &str) -> anyhow::Result<(TuiCommandItem, TuiKeyBinding)> {
    let pos = value
        .find('=')
        .ok_or_else(|| anyhow!("invalid binding value: expected format `item=value`"))?;
    let item = TuiCommandItem::try_from(&value[..pos])?;
    let binding = TuiKeyBinding::try_from(&value[pos + 1..])?;
    if item == TuiCommandItem::DeprecatedTogglePrivacy {
        return Err(anyhow!(
            "toggle-privacy is deprecated, use expand-privacy and contract-privacy instead"
        ));
    }
    Ok((item, binding))
}

fn parse_duration(value: &str) -> anyhow::Result<Duration> {
    Ok(humantime::parse_duration(value)?)
}

fn parse_addr(value: &str) -> anyhow::Result<IpAddr> {
    Ok(IpAddr::from_str(value)?)
}
