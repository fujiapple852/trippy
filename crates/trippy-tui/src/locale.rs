use itertools::Itertools;
use serde::Deserialize;
use std::cell::RefCell;
use std::collections::HashMap;
use std::str::FromStr;
use std::sync::OnceLock;
use unic_langid::LanguageIdentifier;

const FALLBACK_LOCALE: &str = "en";

/// Set the locale for the application.
///
/// If the given locale is `None` the system locale is tried. If the system locale cannot be
/// determined then the fallback locale is used.
///
/// In all cases, the language part of the locale is used if the full locale is not supported.
pub fn set_locale(locale: Option<&str>) -> String {
    let new_locale = calculate_locale(locale, sys_locale::get_locale().as_deref());
    store_locale(&new_locale);
    new_locale
}

/// Get the available locales.
pub fn available_locales() -> Vec<&'static str> {
    data()
        .0
        .iter()
        .flat_map(|(_, v)| v.0.keys().map(AsRef::as_ref))
        .unique()
        .sorted_unstable()
        .collect::<Vec<_>>()
}

/// A macro to translate an item to the current locale.
#[macro_export]
macro_rules! t {
    ($key:expr) => {
        std::borrow::Cow::Borrowed($crate::locale::__translate($key))
    };
    ($key:expr, $($kt:ident = $kv:expr),+) => {
        {
            let string = t!($key);
            $(
                let string = string.replace(concat!("%{", stringify!($kt), "}"), &$kv.to_string());
            )+
            string
        }
    };
    ($key:expr, $($kt:literal => $kv:expr),+) => {
        {
            let string = t!($key);
            $(
                let string = string.replace(concat!("%{", $kt, "}"), &$kv.to_string());
            )+
            string
        }
    };
}

/// Translate an item to the current locale.
///
/// This function is public as it is used by the `t!` macro, however is not considered part of the
/// public interface.
#[doc(hidden)]
pub fn __translate(item: &str) -> &str {
    let locale = CURRENT_LOCALE.with(Clone::clone);
    let binding = locale.borrow();
    translate_locale(item, binding.as_str())
}

/// Translate an item to a specific locale.
///
/// If the item does not exists, the key is returned. Otherwise, if item does not contain the
/// locale, the fallback locale is used. If the fallback locale does not exist, the key is
/// returned.
fn translate_locale<'a>(item: &'a str, locale: &str) -> &'a str {
    if let Some(key) = data().0.get(item) {
        if let Some(value) = key.0.get(locale) {
            value
        } else if let Some(value) = key.0.get(FALLBACK_LOCALE) {
            value
        } else {
            item
        }
    } else {
        item
    }
}

/// Get the locale data.
fn data() -> &'static Data {
    static DATA: OnceLock<Data> = OnceLock::new();
    DATA.get_or_init(|| {
        toml::from_str(include_str!("../locales.toml")).expect("Failed to parse locales.toml")
    })
}

/// This is a map of a item name (i.e. `title_hops`, `awaiting_data`, etc.) to the locale `Item`.
#[derive(Debug, Deserialize)]
struct Data(HashMap<String, Item>);

/// This is a map of locale keys (i.e. `en`, `zh`, etc.) to the translated value.
#[derive(Debug, Deserialize)]
struct Item(HashMap<String, String>);

/// calculate the locale to use.
fn calculate_locale(cfg_locale: Option<&str>, sys_locale: Option<&str>) -> String {
    let preferred = cfg_locale.or(sys_locale).unwrap_or(FALLBACK_LOCALE);
    let locales = available_locales();
    locales
        .contains(&preferred)
        .then(|| preferred.to_string())
        .or_else(|| {
            LanguageIdentifier::from_str(preferred).ok().and_then(|id| {
                let lang = id.language.to_string();
                id.region
                    .map(|r| format!("{lang}-{r}"))
                    .filter(|s| locales.contains(&s.as_str()))
                    .or_else(|| locales.contains(&lang.as_str()).then_some(lang))
            })
        })
        .unwrap_or_else(|| FALLBACK_LOCALE.to_string())
}

thread_local! {
    static CURRENT_LOCALE: RefCell<String> = RefCell::new(String::from(FALLBACK_LOCALE));
}

fn store_locale(new_locale: &str) {
    CURRENT_LOCALE.with(|locale| *locale.borrow_mut() = String::from(new_locale));
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_case::test_case;

    #[test_case(None, None, "en"; "no_locale")]
    #[test_case(Some("en"), None, "en"; "cfg_locale")]
    #[test_case(None, Some("en"), "en"; "sys_locale")]
    #[test_case(Some("en"), Some("en"), "en"; "both_locales")]
    #[test_case(Some("en"), Some("zh"), "en"; "both_locales_mismatch")]
    #[test_case(Some("zh"), Some("en"), "zh"; "both_locales_mismatch_reverse")]
    #[test_case(Some("en-US"), None, "en"; "cfg_locale_dash")]
    #[test_case(None, Some("en-US"), "en"; "sys_locale_dash")]
    #[test_case(Some("en-US"), Some("en-US"), "en"; "both_locales_dash")]
    #[test_case(Some("en-US"), Some("zh-CN"), "en"; "both_locales_mismatch_dash")]
    #[test_case(Some("zh-CN"), Some("en-US"), "zh"; "both_locales_mismatch_reverse_dash")]
    #[test_case(Some("en_US"), None, "en"; "cfg_locale_underscore")]
    #[test_case(None, Some("en_US"), "en"; "sys_locale_underscore")]
    #[test_case(Some("xx"), None, "en"; "cfg_locale_unknown")]
    #[test_case(None, Some("xx"), "en"; "sys_locale_unknown")]
    #[test_case(Some("xx"), Some("xx"), "en"; "both_locales_unknown")]
    #[test_case(Some("en-"), None, "en"; "cfg_locale_invalid_dash")]
    #[test_case(Some("en_"), None, "en"; "cfg_locale_invalid_underscore")]
    #[test_case(Some("en?"), None, "en"; "cfg_locale_invalid_accepted")]
    #[test_case(Some("zh-Hant-TW"), None, "zh-TW"; "cfg_locale_ignore_script")]
    #[test_case(None, Some("zh-Hant-TW"), "zh-TW"; "sys_locale_ignore_script")]
    #[test_case(Some("zh-Hant-TW"), Some("zh-Hant-TW"), "zh-TW"; "both_locales_ignore_script")]
    fn test_set_locale(cfg_locale: Option<&str>, sys_locale: Option<&str>, expected: &str) {
        assert_eq!(calculate_locale(cfg_locale, sys_locale), expected);
    }

    #[test]
    fn test_available_languages() {
        assert_eq!(
            available_locales(),
            vec!["de", "en", "es", "fr", "it", "pt", "ru", "sv", "tr", "zh", "zh-TW"]
        );
    }

    #[test]
    fn test_data_deserialize() {
        assert!(!data().0.is_empty());
    }

    #[test]
    fn test_translate() {
        assert_eq!(translate_locale("title_hops", "en"), "Hops");
        assert_eq!(translate_locale("title_hops", "zh"), "跳");
        assert_eq!(translate_locale("title_hops", "zh-TW"), "跳");
        assert_eq!(translate_locale("unknown_item", "en"), "unknown_item");
        assert_eq!(translate_locale("unknown_locale", "xx"), "unknown_locale");
    }

    #[test]
    fn test_translate_macro() {
        assert_eq!(t!("title_hops"), "Hops");
        assert_eq!(t!("awaiting_data"), "Awaiting data...");
        assert_eq!(t!("unknown_item"), "unknown_item");
    }

    #[test]
    fn test_zh_tw_translations() {
        // Test key Traditional Chinese translations
        assert_eq!(translate_locale("auto", "zh-TW"), "自動");
        assert_eq!(translate_locale("status_failed", "zh-TW"), "失敗");
        assert_eq!(translate_locale("status_running", "zh-TW"), "執行中");
        assert_eq!(translate_locale("title_settings", "zh-TW"), "設定");
        assert_eq!(translate_locale("help_tagline", "zh-TW"), "網路診斷工具");
        assert_eq!(translate_locale("column_loss_pct", "zh-TW"), "封包遺失率");
        assert_eq!(translate_locale("rtt", "zh-TW"), "往返時間");
    }
}
