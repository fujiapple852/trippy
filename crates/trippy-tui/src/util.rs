#[cfg(test)]
pub fn insta<F: FnOnce()>(name: &str, f: F) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_suffix(name.replace(' ', "_"));
    settings.set_snapshot_path("../tests/resources/snapshots");
    settings.set_omit_expression(true);
    settings.bind(f);
}

#[cfg(test)]
pub fn remove_whitespace(mut s: String) -> String {
    s.retain(|c| !c.is_whitespace());
    s
}
