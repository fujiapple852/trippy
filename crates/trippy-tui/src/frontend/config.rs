use crate::config::{AddressMode, AsMode, GeoIpMode, TuiColumns, TuiTheme};
use crate::config::{IcmpExtensionMode, TuiBindings};
use crate::frontend::binding::Bindings;
use crate::frontend::columns::Columns;
use crate::frontend::theme::Theme;
use chrono_tz::Tz;
use std::time::Duration;

/// Tui configuration.
#[derive(Debug)]
pub struct TuiConfig {
    /// Refresh rate.
    pub refresh_rate: Duration,
    /// The maximum ttl of hops which will be masked for privacy.
    pub privacy_max_ttl: Option<u8>,
    /// Preserve screen on exit.
    pub preserve_screen: bool,
    /// How to render addresses.
    pub address_mode: AddressMode,
    /// Lookup autonomous system (AS) information.
    pub lookup_as_info: bool,
    /// How to render autonomous system (AS) data.
    pub as_mode: AsMode,
    /// How to render ICMP extensions.
    pub icmp_extension_mode: IcmpExtensionMode,
    /// How to render `GeoIp` data.
    pub geoip_mode: GeoIpMode,
    /// The maximum number of addresses to show per hop.
    pub max_addrs: Option<u8>,
    /// The Tui color theme.
    pub theme: Theme,
    /// The Tui keyboard bindings.
    pub bindings: Bindings,
    /// The columns to display in the hops table.
    pub tui_columns: Columns,
    pub geoip_mmdb_file: Option<String>,
    pub dns_resolve_all: bool,
    /// The current locale.
    pub locale: String,
    pub timezone: Option<Tz>,
}

impl TuiConfig {
    #[expect(clippy::too_many_arguments)]
    pub fn new(
        refresh_rate: Duration,
        privacy_max_ttl: Option<u8>,
        preserve_screen: bool,
        address_mode: AddressMode,
        lookup_as_info: bool,
        as_mode: AsMode,
        icmp_extension_mode: IcmpExtensionMode,
        geoip_mode: GeoIpMode,
        max_addrs: Option<u8>,
        tui_theme: TuiTheme,
        tui_bindings: &TuiBindings,
        tui_columns: &TuiColumns,
        geoip_mmdb_file: Option<String>,
        dns_resolve_all: bool,
        locale: String,
        timezone: Option<Tz>,
    ) -> Self {
        Self {
            refresh_rate,
            privacy_max_ttl,
            preserve_screen,
            address_mode,
            lookup_as_info,
            as_mode,
            icmp_extension_mode,
            geoip_mode,
            max_addrs,
            theme: Theme::from(tui_theme),
            bindings: Bindings::from(*tui_bindings),
            tui_columns: Columns::from(tui_columns.clone()),
            geoip_mmdb_file,
            dns_resolve_all,
            locale,
            timezone,
        }
    }
}
