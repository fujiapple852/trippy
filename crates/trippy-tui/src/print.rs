use crate::config::{Args, TuiCommandItem, TuiThemeItem};
use crate::locale::available_locales;
use clap::CommandFactory;
use clap_complete::Shell;
use std::process;
use strum::VariantNames;

pub fn print_tui_theme_items() {
    println!("{}", tui_theme_items());
    process::exit(0);
}

pub fn print_tui_binding_commands() {
    println!("{}", tui_binding_commands());
    process::exit(0);
}

pub fn print_config_template() {
    println!("{}", include_str!("../trippy-config-sample.toml"));
    process::exit(0);
}

pub fn print_shell_completions(shell: Shell) -> anyhow::Result<()> {
    println!("{}", shell_completions(shell)?);
    process::exit(0);
}

pub fn print_man_page() -> anyhow::Result<()> {
    println!("{}", man_page()?);
    process::exit(0);
}

pub fn print_locales() {
    println!("TUI locales: {}", available_locales().join(", "));
    process::exit(0);
}

fn tui_theme_items() -> String {
    format!(
        "TUI theme color items: {}",
        TuiThemeItem::VARIANTS.join(", ")
    )
}

fn tui_binding_commands() -> String {
    format!(
        "TUI binding commands: {}",
        TuiCommandItem::VARIANTS.join(", ")
    )
}

fn shell_completions(shell: Shell) -> anyhow::Result<String> {
    let mut cmd = Args::command();
    let name = cmd.get_name().to_string();
    let mut buffer: Vec<u8> = vec![];
    clap_complete::generate(shell, &mut cmd, name, &mut buffer);
    Ok(String::from_utf8(buffer)?)
}

fn man_page() -> anyhow::Result<String> {
    let cmd = Args::command();
    let mut buffer: Vec<u8> = vec![];
    clap_mangen::Man::new(cmd).render(&mut buffer)?;
    Ok(String::from_utf8(buffer)?)
}

#[cfg(test)]
pub mod tests {
    use super::*;
    use crate::util::{insta, remove_whitespace};
    use test_case::test_case;

    #[test_case(&tui_theme_items(), "tui theme items match"; "tui theme items match")]
    #[test_case(&tui_binding_commands(), "tui binding commands match"; "tui binding commands match")]
    #[test_case(&shell_completions(Shell::Bash).unwrap(), "generate bash shell completions"; "generate bash shell completions")]
    #[test_case(&shell_completions(Shell::Elvish).unwrap(), "generate elvish shell completions"; "generate elvish shell completions")]
    #[test_case(&shell_completions(Shell::Fish).unwrap(), "generate fish shell completions"; "generate fish shell completions")]
    #[test_case(&shell_completions(Shell::PowerShell).unwrap(), "generate powershell shell completions"; "generate powershell shell completions")]
    #[test_case(&shell_completions(Shell::Zsh).unwrap(), "generate zsh shell completions"; "generate zsh shell completions")]
    #[test_case(&man_page().unwrap(), "generate man page"; "generate man page")]
    fn test_output(actual: &str, name: &str) {
        insta(name, || {
            insta::assert_snapshot!(remove_whitespace(actual.to_string()));
        });
    }
}
