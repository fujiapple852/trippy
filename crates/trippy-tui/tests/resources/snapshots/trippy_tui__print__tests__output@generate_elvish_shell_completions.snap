---
source: crates/trippy-tui/src/print.rs
---
usebuiltin;usestr;setedit:completion:arg-completer[trip]={|@words|fnspaces{|n|builtin:repeat$n''|str:join''}fncand{|textdesc|edit:complex-candidate$text&display=$text''(spaces(-14(wcswidth$text)))$desc}varcommand='trip'forword$words[1..-1]{if(str:has-prefix$word'-'){break}setcommand=$command';'$word}varcompletions=[&'trip'={cand-c'Configfile'cand--config-file'Configfile'cand-m'Outputmode[default:tui]'cand--mode'Outputmode[default:tui]'cand-p'Tracingprotocol[default:icmp]'cand--protocol'Tracingprotocol[default:icmp]'cand-F'Theaddressfamily[default:system]'cand--addr-family'Theaddressfamily[default:system]'cand-P'Thetargetport(TCP&UDPonly)[default:80]'cand--target-port'Thetargetport(TCP&UDPonly)[default:80]'cand-S'Thesourceport(TCP&UDPonly)[default:auto]'cand--source-port'Thesourceport(TCP&UDPonly)[default:auto]'cand-A'ThesourceIPaddress[default:auto]'cand--source-address'ThesourceIPaddress[default:auto]'cand-I'Thenetworkinterface[default:auto]'cand--interface'Thenetworkinterface[default:auto]'cand-i'Theminimumdurationofeveryround[default:1s]'cand--min-round-duration'Theminimumdurationofeveryround[default:1s]'cand-T'Themaximumdurationofeveryround[default:1s]'cand--max-round-duration'Themaximumdurationofeveryround[default:1s]'cand-g'TheperiodoftimetowaitforadditionalICMPresponsesafterthetargethasresponded[default:100ms]'cand--grace-duration'TheperiodoftimetowaitforadditionalICMPresponsesafterthetargethasresponded[default:100ms]'cand--initial-sequence'Theinitialsequencenumber[default:33434]'cand-R'TheEqual-costMulti-Pathroutingstrategy(UDPonly)[default:classic]'cand--multipath-strategy'TheEqual-costMulti-Pathroutingstrategy(UDPonly)[default:classic]'cand-U'Themaximumnumberofin-flightICMPechorequests[default:24]'cand--max-inflight'Themaximumnumberofin-flightICMPechorequests[default:24]'cand-f'TheTTLtostartfrom[default:1]'cand--first-ttl'TheTTLtostartfrom[default:1]'cand-t'ThemaximumnumberofTTLhops[default:64]'cand--max-ttl'ThemaximumnumberofTTLhops[default:64]'cand--packet-size'ThesizeofIPpackettosend(IPheader+ICMPheader+payload)[default:84]'cand--payload-pattern'TherepeatingpatterninthepayloadoftheICMPpacket[default:0]'cand-Q'TheTOS(i.e.DSCP+ECN)IPheadervalue(IPv4only)[default:0]'cand--tos'TheTOS(i.e.DSCP+ECN)IPheadervalue(IPv4only)[default:0]'cand--read-timeout'Thesocketreadtimeout[default:10ms]'cand-r'HowtoperformDNSqueries[default:system]'cand--dns-resolve-method'HowtoperformDNSqueries[default:system]'cand--dns-timeout'ThemaximumtimetowaittoperformDNSqueries[default:5s]'cand--dns-ttl'Thetime-to-live(TTL)ofDNSentries[default:300s]'cand-s'Themaximumnumberofsamplestorecordperhop[default:256]'cand--max-samples'Themaximumnumberofsamplestorecordperhop[default:256]'cand--max-flows'Themaximumnumberofflowstorecord[default:64]'cand-a'Howtorenderaddresses[default:host]'cand--tui-address-mode'Howtorenderaddresses[default:host]'cand--tui-as-mode'Howtorenderautonomoussystem(AS)information[default:asn]'cand--tui-custom-columns'CustomcolumnstobedisplayedintheTUIhopstable[default:holsravbwdt]'cand--tui-icmp-extension-mode'HowtorenderICMPextensions[default:off]'cand--tui-geoip-mode'HowtorenderGeoIpinformation[default:short]'cand-M'Themaximumnumberofaddressestoshowperhop[default:auto]'cand--tui-max-addrs'Themaximumnumberofaddressestoshowperhop[default:auto]'cand--tui-refresh-rate'TheTUIrefreshrate[default:100ms]'cand--tui-privacy-max-ttl'Themaximumttlofhopswhichwillbemaskedforprivacy[default:none]'cand--tui-locale'ThelocaletousefortheTUI[default:auto]'cand--tui-timezone'ThetimezonetousefortheTUI[default:auto]'cand--tui-theme-colors'TheTUIthemecolors[item=color,item=color,..]'cand--tui-key-bindings'TheTUIkeybindings[command=key,command=key,..]'cand-C'Thenumberofreportcyclestorun[default:10]'cand--report-cycles'Thenumberofreportcyclestorun[default:10]'cand-G'ThesupportedMaxMindorIPinfoGeoIpmmdbfile'cand--geoip-mmdb-file'ThesupportedMaxMindorIPinfoGeoIpmmdbfile'cand--generate'Generateshellcompletion'cand--log-format'Thedebuglogformat[default:pretty]'cand--log-filter'Thedebuglogfilter[default:trippy=debug]'cand--log-span-events'Thedebuglogformat[default:off]'cand-u'Tracewithoutrequiringelevatedprivilegesonsupportedplatforms[default:false]'cand--unprivileged'Tracewithoutrequiringelevatedprivilegesonsupportedplatforms[default:false]'cand--udp'TraceusingtheUDPprotocol'cand--tcp'TraceusingtheTCPprotocol'cand--icmp'TraceusingtheICMPprotocol'cand-4'UseIPv4only'cand--ipv4'UseIPv4only'cand-6'UseIPv6only'cand--ipv6'UseIPv6only'cand-e'ParseICMPextensions'cand--icmp-extensions'ParseICMPextensions'cand-y'TracetoallIPsresolvedfromDNSlookup[default:false]'cand--dns-resolve-all'TracetoallIPsresolvedfromDNSlookup[default:false]'cand-z'Lookupautonomoussystem(AS)informationduringDNSqueries[default:false]'cand--dns-lookup-as-info'Lookupautonomoussystem(AS)informationduringDNSqueries[default:false]'cand--tui-preserve-screen'Preservethescreenonexit[default:false]'cand--print-tui-theme-items'PrintallTUIthemeitemsandexit'cand--print-tui-binding-commands'PrintallTUIcommandsthatcanbeboundandexit'cand--generate-man'GenerateROFFmanpage'cand--print-config-template'Printatemplatetomlconfigfileandexit'cand--print-locales'PrintallavailableTUIlocalesandexit'cand-v'Enableverbosedebuglogging'cand--verbose'Enableverbosedebuglogging'cand-h'Printhelp(seemorewith''--help'')'cand--help'Printhelp(seemorewith''--help'')'cand-V'Printversion'cand--version'Printversion'}]$completions[$command]}
